operators/cmd/tenant-operator/main.go

```diff
@@ -61,30 +77,80 @@ func main() {
 		LeaderElectionID:   "f547a6ba.crownlabs.polito.it",
 	})
 	if err != nil {
-		setupLog.Error(err, "unable to start manager")
+		klog.Error("Unable to start manager", err)
 		os.Exit(1)
 	}
 
+	kcClient, kcToken := prepareKcClient(kcURL, kcAdminUser, kcAdminPsw)
+
+	go checkAndRenewTokenPeriodically(context.Background(), kcClient, kcToken, kcAdminUser, kcAdminPsw, 2*time.Minute, 5*time.Minute)
+
 	if err = (&controllers.TenantReconciler{
 		Client: mgr.GetClient(),
 		Scheme: mgr.GetScheme(),
 	}).SetupWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create controller", "controller", "Tenant")
+		klog.Error("Unable to create controller for Tenant", err)
 		os.Exit(1)
 	}
 	if err = (&controllers.WorkspaceReconciler{
-		Client: mgr.GetClient(),
-		Scheme: mgr.GetScheme(),
+		Client:         mgr.GetClient(),
+		Scheme:         mgr.GetScheme(),
+		KcClient:       kcClient,
+		KcToken:        kcToken,
+		KcTargetRealm:  kcTargetRealm,
+		KcTargetClient: kcTargetClient,
 	}).SetupWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create controller", "controller", "Workspace")
+		klog.Error("Unable to create controller for Workspace", err)
 		os.Exit(1)
 	}
 	// +kubebuilder:scaffold:builder
 
-	setupLog.Info("starting manager")
+	klog.Info("Starting manager")
 	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
-		setupLog.Error(err, "problem running manager")
+		klog.Error("Problem running manager", err)
 		os.Exit(1)
 	}
 
 }
+
+// prepareKcClient sets up a keycloak client with the specififed parameters and performs the first login
+func prepareKcClient(kcURL, kcAdminUser, kcAdminPsw string) (gocloak.GoCloak, *gocloak.JWT) {
+
+	kcClient := gocloak.NewClient(kcURL)
+	token, kcErr := kcClient.LoginAdmin(context.Background(), kcAdminUser, kcAdminPsw, "master")
```

Comment:
I would avoid to hardcode "master" as a parameter. As it is likely not to change (and to avoid creating a parameter), you could just create a constant and add a bit of description there

Reply:
Ok, if I make it a parameter is a problem? I would do it that way for consistency and not having to touch the code in need for that change

Reply:
Definitely even better